Add ability to load tasks/handlers from dir

desa opened the pull request:
# Todo

- [x] Add batch tasks to example of mock directories.

# Done
- [x] Load files from directory
- [x] Extend API to allow for new tickscript definition
- [x] Update Tickscript definition
- [x] Update Template var defition
- [x] Update Handler Definition

# Todo
- [x] Add tests
- [x] Reload directory on OS signal
- [x] Clean up

###### Required for all non-trivial PRs
- [x] Rebased/mergable
- [x] Tests pass
- [x] CHANGELOG.md updated

desa commented on tick/ast/node.go:
```diff
@@ -361,6 +362,49 @@ func (n *BinaryNode) Equal(o interface{}) bool {
+	}
+}
+
+// TODO: Use Better typing eventually
```
remove

desa commented on tick/ast/node.go:
```diff
@@ -981,6 +1025,100 @@ func newProgram(p position) *ProgramNode {
+	return dbrps
+}
+
+// TODO: better type here
```
remove

desa commented on tick/ast/node.go:
```diff
@@ -981,6 +1025,100 @@ func newProgram(p position) *ProgramNode {
+			if cn, ok := nn.(*DeclarationNode).Right.(*ChainNode); ok {
+				// TODO: dont repeat logic
+				var n = cn.Left
+			Loop1:
```
rename DeclLoop

desa commented on tick/ast/node.go:
```diff
@@ -981,6 +1025,100 @@ func newProgram(p position) *ProgramNode {
+			}
+		case *ChainNode:
+			var n = nn.(*ChainNode).Left
+		Loop2:
```
rename ChainLoop

desa commented on tick/ast/parser.go:
```diff
@@ -237,12 +237,25 @@ func (p *parser) statement() Node {
 	}
 }
 
 //parse a declaration statement
```
parse dbrp statement

desa commented on tick/ast/parser_test.go:
```diff
@@ -127,6 +127,42 @@ func TestParseStatements(t *testing.T) {
 		Root   Node
 		err    error
 	}{
+		// TODO: Allow use of " in dbrp declarations
```
remove

desa commented on services/load/service.go:
```diff
@@ -0,0 +1,404 @@
+
+// TaskFiles gets a slice of all files with the .tick file extension
+// in the configured task directory.
+func (s *Service) TaskFiles() (tickscripts []string, err error) {
```
unexport

desa commented on services/load/service.go:
```diff
@@ -0,0 +1,404 @@
+// TemplateFiles gets a slice of all files with the .tick file extension
+// and any associated files with .json, .yml, and .yaml file extentions
+// in the configured template directory.
+func (s *Service) TemplateFiles() (tickscripts []string, tmplVars []string, err error) {
```
unexport

desa commented on services/load/service.go:
```diff
@@ -0,0 +1,404 @@
+
+// HandlerFiles gets a slice of all files with the .json, .yml, and
+// .yaml file extentions in the configured handler directory.
+func (s *Service) HandlerFiles() ([]string, error) {
```
unexport

desa commented on services/load/service.go:
```diff
@@ -0,0 +1,404 @@
+	return handlers, nil
+}
+
+func (s *Service) Load() error {
```
Add logging

desa commented on services/load/service.go:
```diff
@@ -0,0 +1,404 @@
+	return nil
+}
+
+func (s *Service) loadTickscripts() error {
```
change to load tasks

desa commented on services/load/service.go:
```diff
@@ -0,0 +1,404 @@
+	return nil
+}
+
+func (s *Service) loadTickscript(f string) error {
```
ditto

desa commented on services/load/service.go:
```diff
@@ -0,0 +1,404 @@
+}
+
+func (s *Service) loadTemplate(f string) error {
+	file, err := os.Open(f)
```
`defer file.Close()`

desa commented on services/task_store/service.go:
```diff
@@ -865,14 +903,50 @@ func (ts *Service) handleUpdateTask(w http.ResponseWriter, r *http.Request) {
 			updated.Type = BatchTask
 		}
 
+		oldPn, err := ast.NewProgramNodeFromTickscript(updated.TICKscript)
```
Cant actually do it cleanly.

desa commented:
@nathanielc This should be in a reviewable state. I still need to add a few more tests to the server package so that I can be completely confident in the changes, but it should reviewable.

desa commented on task.go:
```diff
@@ -18,6 +18,7 @@ type TaskType int
 const (
 	StreamTask TaskType = iota
 	BatchTask
+	InvalidTask
```
yes

nathanielc commented on tick/ast/node.go:
```diff
@@ -981,6 +1024,101 @@ func newProgram(p position) *ProgramNode {
+}
+
+// TODO: Not sure how I feel about client.TaskType importing from client
+func (n *ProgramNode) DBRPs() []client.DBRP {
```
In general importing the client types in the server code makes sense as it ensures that the server and client are using the same types. But here we are not server code but rather an independent package that handles parsing scripts.

I recommend that you use the AST nodes as the public API from the `tick` package and move these methods into the server code where needed.  So it would look something like `dbrpsFromProgram(*ast.ProgramNode) []client.DBRP`. Does that help?

nathanielc commented on tick/ast/node.go:
```diff
@@ -981,6 +1024,101 @@ func newProgram(p position) *ProgramNode {
 	}
 }
 
+func NewProgramNodeFromTickscript(tickscript string) (*ProgramNode, error) {
```
I don't really like this function name, but I get that it is a common need to be able to type assert the ast.Node to its concrete type. Should we just have the `Parse` method return the concrete type?

nathanielc replied:
FWIW The reasons I don't like the name is without reading the function's source I thought it was doing some kind of special operation to return a ProgramNode from a TICKscript instead of just using the Parse method.

Another option would be to move this method to the package where it is consumed as a helper method.

desa commented on services/load/service.go:
```diff
@@ -0,0 +1,416 @@
+	"github.com/pkg/errors"
+)
+
+var defaultURL = "http://localhost:9092"
```
Yeah I really like that.

desa commented on client/v1/client.go:
```diff
@@ -2282,3 +2289,107 @@ func (d *Duration) UnmarshalText(data []byte) error {
 	return nil
 }
+
+// TODO: this is also used in kapacitor/main.go
```
Old comment. Previously this code was in `kapacitor/main.go` and unexported. I moved it here and exported it. The `TODO` item was meant to remind me to update `kapacitor/main.go` to use this code instead.

desa commented on cmd/kapacitor/main.go:
```diff
@@ -822,7 +804,7 @@ func doDefineTemplate(args []string) error {
 
 func defineTopicHandlerUsage() {
-	var u = `Usage: kapacitor define-topic-handler <topic id> <handler id> <path to handler spec file>
+	var u = `Usage: kapacitor define-topic-handler <path to handler spec file>
```
Yup.

desa commented on cmd/kapacitord/main.go:
```diff
@@ -89,15 +89,31 @@ func (m *Main) Run(args ...string) error {
+			case <-reloadSignal:
+				m.Logger.Println("I! SIGHUP received, Reloading tasks/templates/directory directory...")
+				if err := cmd.Server.LoadService.Load(); err != nil {
+					// TODO: should this be a sufficient condition to shutdown the server?
```
> What kinds of errors can LoadService.Load return?

Errors are either from reading files or making requests to the API.

> I would prefer the operation to be atomic, meaning if it fails it logs the failure and then reverts to its previous state.
> That is not really possible since you are going through the API instead of calling into the DB services directly.

I think this might be possible, albeit painful. And I'd have to make some assumptions.
I could figure out what tasks, templates, and handlers would result from a new `Load` operation, save their state temporarily, and then, in the case of a failure, rewrite the old values back through the API.

> I am not sure what we should do here. Thoughts?

I'm very much in the same boat.  The three possibilities I see

1. Best effort, don't crash the server, log the errors
2. Crash the server and treat the files like configuration
3. Don't crash the server, restore to a previous state

I'm inclined to lean towards 2 or 3, but I think that it might be possible to do things to minimize the likelihood of an error.

For example, we could add a `dry-run` query param to the API that attempts to do everything that the regular request to the API would, but without actually committing any changes? I think this should mostly be feasible (although possibly not desirable).

This way we could do a dry run to "validate" the files that are currently there. Or something like this. Essentially a best effort to validate that the entire `Load` operation will be successful, without touching the API.

Alternatively, I think the case could be for 2, in that these files should be treated like configuration and bad configuration should stop the runtime.

desa commented on cmd/kapacitor/main.go:
```diff
@@ -687,29 +629,69 @@ func doDefine(args []string) error {
 			Status:     client.Disabled,
-		})
+		}
+		if *dfile != "" {
```
Not sure I totally follow. Are you saying to do 

```go

		if *dfile != "" {
			o, err := fileVars.CreateTaskOptions()
			if err != nil {
				return err
			}
		_, err = cli.CreateTask(o)
		} else {
		o := client.CreateTaskOptions{
			ID:         id,
			TemplateID: *dtemplate,
			Type:       ttype,
			DBRPs:      ddbrp,
			TICKscript: script,
			Vars:       vars,
			Status:     client.Disabled,
		}
		_, err = cli.CreateTask(o)
            }
```

this instead?

desa commented on cmd/kapacitord/main.go:
```diff
@@ -89,15 +89,31 @@ func (m *Main) Run(args ...string) error {
 		signalCh := make(chan os.Signal, 1)
 		signal.Notify(signalCh, os.Interrupt, syscall.SIGTERM)
+		reloadSignal := make(chan os.Signal, 1)
+		signal.Notify(reloadSignal, os.Interrupt, syscall.SIGHUP)
```
I think we'll still need the loop since the `SIGHUP` doesn't necessarily crash the runtime anymore.

desa commented on server/server.go:
```diff
@@ -296,6 +302,8 @@ func (s *Server) appendStorageService() {
 	s.StorageService = srv
 	s.AppendService("storage", srv)
+
+	return
```
Accident

desa commented on services/load/service.go:
```diff
@@ -0,0 +1,416 @@
+
+	handlersDir := s.config.handlersDir()
+
+	files, err := ioutil.ReadDir(handlersDir)
```
in `load/config.go` the `Validate` method requires that the directory specified must have subdirectories `tasks`, `templates`, and `handlers`. Should I relax this requirement?

desa commented on cmd/kapacitord/main.go:
```diff
@@ -89,33 +89,42 @@ func (m *Main) Run(args ...string) error {
-						}()
-						break Loop
+
+			case s := <-signalCh:
```
@nathanielc there has to be a better way than this, but I couldn't find an example. This could be my end-of-day brain not thinking straight. Thoughts?

desa commented:
https://github.com/influxdata/chronograf/pull/1721 will be able to remove the `stream/batch` button from tickscript UI when this is merged

desa commented:
@121watts ^^

desa commented:
Seems to be having an issue with batch tasks. Not exactly sure why. Will investigate.

F4ncyMooN commented:
Hi, I'm looking for a convergence method for alerts. 
For example, I want to monitor on 100 hosts which are located in 2 different DCs(50 in DC1, 50 in DC2), if 90% of DC1's hosts failed, only DC's alerts will be delivered.
I think this should be implemented by myself in some handlers.
Is it possible if I create some handlers without change the source code of kapacitor(just like udf) after this PR?

desa commented:
@F4ncyMooN This PR is a little bit different. This PR introduces a way to load tasks and templates from a directory.

nathanielc commented on cmd/kapacitor/main.go:
```diff
@@ -542,72 +541,15 @@ var (
-
-// Parse string of the form "db"."rp" where the quotes are optional but can include escaped quotes
-// within the strings.
-func (d *dbrps) Set(value string) error {
```
Why was this moved to the client package? The `Set` method is part of the `flag.Value` interface, which seems like it belongs here in the main package.

desa replied:
`client.TaskVars` ended up needing the same DBRP logic so rather than having two copies of the type, one in the `main` package and one in the `client` package I figured that I'd move it into client. I'm happy to have two separate types if you think thats best.

nathanielc replied:
Why did the `client.TaskVars` need to implement the `flag.Value` interface?

nathanielc replied:
Maybe the type in the client package should implement a TextUnmarshaler interface and then the type in the main package uses the type from the client package and adds the Set method to its own type? Does that work?

desa replied:
Yeah I like that.

desa replied:
@nathanielc Looks like implementing `TextUnmarshaler` causes issues with the regular JSON marshaling. Does having a method `Unmarshal` on the `client.DBRP` type work?

nathanielc commented on cmd/kapacitord/main.go:
```diff
@@ -88,16 +89,44 @@ func (m *Main) Run(args ...string) error {
-			}()
+	Loop:
+		for {
+			select {
```
This select statement has only a single case, which means this can be simplified to `for s := range signalCh {...}`.

nathanielc replied:
Hmm, ok so further down we check for a second signal. Maybe we should just keep a state about the signals we have received so its clear how transitions occur.

nathanielc commented on cmd/kapacitord/main.go:
```diff
@@ -88,16 +89,44 @@ func (m *Main) Run(args ...string) error {
+						}
+					}
+
+					// This should never happen
```
What about os.Interrupt?

desa replied:
@nathanielc  what should happen on `os.Interrupt`?

desa replied:
removing comment.

nathanielc commented on cmd/kapacitord/main.go:
```diff
@@ -88,16 +89,44 @@ func (m *Main) Run(args ...string) error {
+					m.Logger.Println("I! SIGHUP received, Reloading tasks/templates/handlers directory...")
+					if err := cmd.Server.LoadService.Load(); err != nil {
+						m.Logger.Println(fmt.Sprintf("E! Failed to reload tasks/templates/handlers: %s", err))
+						if _, ok := err.(load.HardError); ok {
```
What other kinds of errors are returned? Should we push this logic into the server type? Maybe something in this method https://github.com/influxdata/kapacitor/blob/master/server/server.go#L882. Plus a method on the Server type to reload.

desa replied:
Not totally sure I follow. I think this logic might belong here. `Load` can return few different error types, but there's a configuration setting that specifies what should happen to the runtime in the case of an error. If the configuration says crash on error we wrap any errors coming from `Load` the error with a `load.HardError`.

desa replied:
Fixed

nathanielc commented on cmd/kapacitord/run/command.go:
```diff
@@ -127,6 +127,11 @@ func (cmd *Command) Run(args ...string) error {
 		return fmt.Errorf("open server: %s", err)
 	}
+
+	if err := s.LoadService.Load(); err != nil {
```
Why isn't this part of the Open process for the Server?

desa replied:
fixed.

desa commented:
@nathanielc addressed most of the issues, only thing I was a bit unsure of was https://github.com/influxdata/kapacitor/pull/1481#discussion_r138128327

nathanielc commented on cmd/kapacitord/main.go:
```diff
@@ -87,16 +88,38 @@ func (m *Main) Run(args ...string) error {
+
+			case syscall.SIGHUP.String():
+				m.Diag.Info("SIGHUP received, reloading tasks/templates/handlers directory...")
+				if err := cmd.Server.LoadService.Load(); err != nil {
```
I was thinking, instead of doing cmd.Server.LoadService.Load() and handing the error types here in the main package, create a method on the server called Reload or something, that if it returns an error at all causes the Server to crash.
Then inside that method determine if the underlying errors returned from LoadService.Load are work crashing. That way the main function crashes if it gets any error, and the Server contains the logic on what errors cause crashes.

Does that help?

desa replied:
yup.

desa commented:
@nathanielc I think I addressed all the previous issues.

nathanielc commented:
I am testing this out locally, here are some of the issues I ran into:

* The `kapacitor config` command produces a load section with an empty directory, it would be good to apply a default dir like we do for the other dirs. See https://github.com/influxdata/kapacitor/blob/master/server/config.go#L181-L184
* After configuring a directory Kapacitor now crashes on startup because the `<dir>/tasks` dir doesn't exist. Kapacitor should treat a lack of dir as there being no tasks configured, not as a condition to crash the server
* The `hard` config option seems to not do anything different whether its on or off. Specifically I had a tick in the dir with some invalid syntax. NO matter what I could not get kapacitor to start without crashing.
* When defining a handler I followed the example [here](https://github.com/influxdata/kapacitor/blob/dbd0d971a614ee563182e95ebf2b29b7d55d0365/examples/load/README.md#handlers) and I got an error since the handler ID was missing from the file.
* I changed the ID of a handler and the old handler is still around, it was not deleted. I expected the old handler to be removed
* I renamed a task and now both the old and new tasks are defined. I expected the old task to be deleted.
* At first I could not figure out how to create a task from a template. I originally put it in the `tasks` directory since its a task and just refers back to the template. Once I move the yaml file back to the template I got it working. I think tasks should exist in the `tasks` dir even if they refer to a template.
* I expected an error if I forgot to define a `template-id` in a template task that already exists, but it doesn't error. It works, which means the files can drift from the real state since its possible to edit a file that would cause an error if created new, but otherwise not.

Those are the issues I ran into, mostly usability issues and the fact the the README in the examples dir is out of date so it lead me astray. 

Its super nice to be able to define the tasks from dirs like this! Thanks!

desa commented:
I am testing this out locally, here are some of the issues I ran into:

> The kapacitor config command produces a load section with an empty directory, it would be good to apply a default dir like we do for the other dirs. See ?>https://github.com/influxdata/kapacitor/blob/master/server/config.go#L181-L184

Fixed

> After configuring a directory Kapacitor now crashes on startup because the <dir>/tasks dir doesn't exist. Kapacitor should treat a lack of dir as there being no tasks configured, not as a condition to crash the server

fixed. had a negation

> At first I could not figure out how to create a task from a template. I originally put it in the tasks directory since its a task and just refers back to the template. Once I move the yaml file back to the template I got it working. I think tasks should exist in the tasks dir even if they refer to a template.

fixed

desa commented:
@nathanielc 

> The hard config option seems to not do anything different whether its on or off. Specifically I had a tick in the dir with some invalid syntax. NO matter what I could not get kapacitor to start without crashing.

This was intentional. Hard errors only apply when re-loading from a directory when the process is running. The idea here was that if you’re starting launching from a directory that the tasks/templates/handlers should be treated as configuration and therefore the server should not start until they are valid. However if you’re running already and you make a change to your tasks/templates/handlers an error with one of those should not crash your server.

> I changed the ID of a handler and the old handler is still around, it was not deleted. I expected the old handler to be removed
I renamed a task and now both the old and new tasks are defined. I expected the old task to be deleted.

How do we do this? My current thought would be to have some kind of `LoadTasks.lock` file that will contain the tasks the were loading from a directory and then diff against that list. That seem reasonable?

> When defining a handler I followed the example here and I got an error since the handler ID was missing from the file.
> Those are the issues I ran into, mostly usability issues and the fact the the README in the examples dir is out of date so it lead me astray.

Which examples were/are problematic? When I use that directory as my load directory everything seems to work fine.

> I expected an error if I forgot to define a template-id in a template task that already exists, but it doesn't error. It works, which means the files can drift from the real state since its possible to edit a file that would cause an error if created new, but otherwise not.

Should this be done by validating the file before even passing it to the API?

nathanielc commented:
>This was intentional. Hard errors only apply when re-loading when the process is running. The idea here was that if you’re starting launching from a directory that the tasks/templates/handlers should be treated as configuration and therefore the server should not start until they are valid. However if you’re running already and you make a change to your tasks/templates/handlers an error with one of those should not crash your server.


Hmm, I am beginning to think we should just remove the `hard` option and always have the reload log errors and never crash the server. This is how apache's httpd works. This way the processes doesn't assume any ownership of the workflow and lets the operator decided how to manage loading tasks from a dir. We should be however logging errors and incrementing a counter if the reload fails. This way it can be easily monitored.

> How do we do this? My current thought would be to have some kind of LoadTasks.lock file that will contain the tasks the were loading from a directory and then diff against that list. That seem reasonable?

My though was track a list of tasks/templates/handlers that are created from the load service and then when the load service runs, if the file system no longer defines that task/template/handler then delete it. This list should be stored in the kapacitor.db file. It could be an explicit list or simply a new property on the data being stored for each item. Does that make sense?


>Which examples were/are problematic? When I use that directory as my load directory everything seems to work fine.

I'll comment in-line on the issues I had.

> Should this be done by validating the file before even passing it to the API?

No, I'd rather all validation logic live in the API, that way if someone isn't using the client they still get proper validation.

nathanielc commented on examples/load/README.md:
```diff
@@ -0,0 +1,138 @@
+
+`dir` specifies the directory where the definition files exist.
+
+The specified directory should have three subdirectories `tasks`, `templates`, and `handlers`.
```
This line implies that the directories must exist, which is not the case.

nathanielc commented on examples/load/README.md:
```diff
@@ -0,0 +1,138 @@
+The `tasks` directory will contain will contain task tickscripts.
+
+The `templates` directory will contain both template tickscripts and the associated templated task
+definition files (either yaml or json).
```
These lines need to be updated to reflect the new behavior.

nathanielc commented on examples/load/README.md:
```diff
@@ -0,0 +1,138 @@
+* `id` - the file name without the tick extension
+* `type` - determined by introspection of the task (stream, batch)
+* `kind` - task, template. defined using a `set statement`
+* `subscriptions` - defined using the `subscribe` keyword followed by a specified subscription
```
Isn't the syntax `dbrp` now?

nathanielc commented on examples/load/README.md:
````diff
@@ -0,0 +1,138 @@
+  "window": {"type" : "duration", "value" : "1m" },
+  "slack_channel": {"type" : "string", "value" : "#alerts_testing" }
+}
+```
````
This is technically correct yaml but quite non-standard a confusing, this reads more like yaml

```yaml
dbrp:
   - telegraf.autogen
   - telegraf.not_autogen
 template: base_template
 vars: 
   measurement:
    type: string
    value: cpu
   where_filter:
    type: lambda
    value: "\"cpu\" == 'cpu-total'"
   groups:
    type: list
    value:
        - type: string
          value: host
        - type: string
          value: dc
   field:
    type: string
    value : usage_idle
   warn:
    type: lambda
    value: "\"mean\" < 30.0"
   crit:
    type: lambda
    value: "\"mean\" < 10.0"
   window:
    type: duration
    value : 1m
   slack_channel:
    type: string
    value: "#alerts_testing"
```

nathanielc commented on examples/load/README.md:
````diff
@@ -0,0 +1,138 @@
+  channel: '#alerts'
+```
+
+the name of the file will be used as the handler id.
````
This is false, the id is read from the file and the filename is not used.

desa commented:
@nathanielc removed the hard error condition and added the logic for deleting stale tasks/templates/handlers that were loaded from a directory. The todo items I have left are

* Add tests for DAO
* Make edits to README
* Cleanup remove logic a little

desa commented:
@nathanielc added tests. Should be ready for review

nathanielc commented on examples/load/README.md:
```diff
@@ -0,0 +1,155 @@
+
+`dir` specifies the directory where the definition files exist.
+
+The specified directory will check to see if any of three subdirectories `tasks`, `templates`, and `handlers` exists.
```
The configured `dir` may contain any of these sub-directories: `tasks`, `templates`, and `handlers`.

nathanielc commented on examples/load/README.md:
```diff
@@ -0,0 +1,155 @@
+The `tasks` directory will contain will contain task tickscripts and and the associated templated task
+definition files (either yaml or json).
+
+The `templates` directory will contain both template tickscripts
```
The `templates` directory will contain template tickscripts

nathanielc commented on examples/load/README.md:
```diff
@@ -0,0 +1,155 @@
+
+The `templates` directory will contain both template tickscripts
+
+The `handlers` directory will contain will contain all topic handler definitions in yaml or json.
```
The `handlers` directory will contain will contain topic handler definitions in yaml or json.

desa replied:
fixed

nathanielc commented on examples/load/README.md:
````diff
@@ -0,0 +1,155 @@
+```
+dbrp "telegraf"."autogen"
+
+var measurement string
````
This example is actually a template so it can't be placed in the tasks dir.

desa replied:
fixed.

nathanielc commented on examples/load/README.md:
```diff
@@ -0,0 +1,155 @@
+dbrp:
+  - telegraf.autogen
+  - telegraf.not_autogen
+template: base_template
```
This should be `templated-id`

desa replied:
fixed.

nathanielc commented on examples/load/README.md:
```diff
@@ -0,0 +1,155 @@
+
+### Templated Tasks
+
+Template variables may be added as either json or yaml.
```
I think the README should have a simple TICKscript above to demonstrate creating tasks.

Then in this Template section it should first define a TICKscript template. Then show how that template can be referenced from a file in the `tasks` directory.

We want to make it clear that anything in the `tasks` dir will create a task, and anything in the `templates` dir will create a template.

desa replied:
Added a section on defining a template before and the templated tasks now use that template. Think that  should be sufficient? Or did you have something else in mind?

nathanielc commented on server/server.go:
```diff
@@ -896,7 +896,7 @@ func (s *Server) Open() error {
 	}
 
 	if err := s.LoadService.Load(); err != nil {
-		return fmt.Errorf("loading tasks/templates/handlers: %s", err)
```
I think its fine for the server to refuse to start if it has bad configuration, but we do not want it to crash from a simple typo.

desa replied:
not sure exactly what to do with this. Should I keep the error here and do some checking of the type of error, or just leave it as is?

nathanielc replied:
What types of errors would you check for? Seems like if it errors at all it should return the error.

nathanielc commented on server/server.go:
```diff
@@ -1060,7 +1060,7 @@ func (s *Server) writeID(file string, id uuid.UUID) error {
 func (s *Server) Reload() error {
```
If this function can't error anymore we should remove the `error` return value and simplify the caller to not need to handle the error.

desa replied:
fixed.

nathanielc commented on services/load/service.go:
```diff
@@ -0,0 +1,666 @@
+}
+
+func (s *Service) loadedTasks() ([]string, error) {
+	items, err := s.items.List("tasks")
```
We should create constants for the `tasks` , `templates` and `handlers` strings, since they are part of the key for the DOA. A typo would corrupt the data :(.

desa replied:
fixed.

nathanielc commented on services/load/service.go:
```diff
@@ -0,0 +1,666 @@
+
+func NewService(c Config, h http.Handler, d Diagnostic) (*Service, error) {
+	cfg := client.Config{
+		URL: defaultURL,
```
Can we add a unique user agent here so its clear the consumer of the API is this internal client from Load?

desa replied:
fixed

desa commented:
@nathanielc  I think I got the completion right, but I'm not totally sure. I sometimes have issues with it on OSX.

Does this seem right?

https://github.com/influxdata/kapacitor/pull/1481/commits/396ab78e847b94949c973e8655161fd9d5b96854

desa commented:
@nathanielc just rebased. I think things should be good to go now.
